File: gopher/src/rozen/calling/calling.go
package main
import "fmt"
import "time"

/*
   You have to implement a method.

   It should return true if it has been called 4 times or less
   in the last 7 seconds.
*/

var start = time.Now()
var callings = 0

func checkStatus() bool {
    // if elapsed time < 7s since first call, and callings <= 4 times
    // increase callings and return true
    // all else, return false
    t := time.Now()
    elapsed := int(time.Since(start).Seconds())
    if elapsed <= 7 {
        if callings < 4 {
            callings++
            return true
        } else {
            callings++
            return false
        }
    // also if > 7s, reset the start/callings, and return false
    } else {
        // now is new beginning
        start = t
        // this is 1st calling
        callings = 0
        // in such case, false
        return false
    }
}

func main() {
    msg := "A gopher named Desire"
    fmt.Printf("Hello gopher!\n%s\n", msg)

    // true
    fmt.Printf("Status shall be true: %v\n", checkStatus())
    // true
    fmt.Printf("Status shall be true: %v\n", checkStatus())
    // true
    fmt.Printf("Status shall be true: %v\n", checkStatus())
    // true
    fmt.Printf("Status shall be true: %v\n", checkStatus())
    // false
    fmt.Printf("Status shall be false: %v\n", checkStatus())
    // false
    fmt.Printf("Status shall be false: %v\n", checkStatus())
    time.Sleep(8 * time.Second)
    // false
    fmt.Printf("Status shall be false: %v\n", checkStatus())
    // true
    fmt.Printf("Status shall be true: %v\n", checkStatus())
    // true
    fmt.Printf("Status shall be true: %v\n", checkStatus())
    // true
    fmt.Printf("Status shall be true: %v\n", checkStatus())
    // true
    fmt.Printf("Status shall be true: %v\n", checkStatus())
    // false
    fmt.Printf("Status shall be false: %v\n", checkStatus())
    // false
    fmt.Printf("Status shall be false: %v\n", checkStatus())
}

File: cpp/calling.cpp
#include <iostream>
#include <cstdlib>
#include <cstdio>
#include <ctime>
#include <string.h>
#include <unistd.h>
using namespace std;

static time_t call_time = 0;

// check and return the timespan in seconds
size_t check_time_interval()
{
    time_t now = time(NULL);
    return size_t(now - call_time);
}

/*
   You have to implement a method.

   It should return true if it has been called 4 times or less
   in the last 7 seconds.
   */

bool checkStatus()
{
    static int called_times = 0;

    if (check_time_interval() <= 7)
    {
        if (called_times++ < 4)
            return true;
        else
            return false;
    }
    else
    {
        // now reset the time boundary
        call_time = time(NULL);

        // also reset the call times to be the first
        called_times = 1;

        // this time must be true
        return true;
    }

    // here never reached.
}

// To execute C++, please define "int main()"
int main() {
    // init call time
    call_time = time(NULL);

    cout << checkStatus() << endl; // True
    cout << checkStatus() << endl; // True
    cout << checkStatus() << endl; // True
    cout << checkStatus() << endl; // True
    cout << checkStatus() << endl; // False
    cout << checkStatus() << endl; // False
    cout << checkStatus() << endl; // False
    cout << checkStatus() << endl; // False
    sleep(8);
    cout << checkStatus() << endl; // True
    cout << checkStatus() << endl; // True
    cout << checkStatus() << endl; // True
    cout << checkStatus() << endl; // True
    sleep(8);
    cout << checkStatus() << endl; // True
    cout << checkStatus() << endl; // True
    cout << checkStatus() << endl; // True
    cout << checkStatus() << endl; // True
    cout << checkStatus() << endl; // False
    cout << checkStatus() << endl; // False
    cout << checkStatus() << endl; // False
    cout << checkStatus() << endl; // False
    return 0;
}
